src/main/scala/org/broadinstitute/dsde/firecloud/core/ProfileActor.scala

```diff
@@ -0,0 +1,86 @@
+package org.broadinstitute.dsde.firecloud.core
+
+import akka.actor.{Actor, Props}
+import akka.event.Logging
+import org.broadinstitute.dsde.firecloud.core.ProfileActor.UpdateProfile
+import org.broadinstitute.dsde.firecloud.model._
+import org.broadinstitute.dsde.firecloud.service.{UserService, FireCloudRequestBuilding}
+import org.broadinstitute.dsde.firecloud.service.PerRequest.RequestComplete
+import spray.client.pipelining._
+import spray.http.HttpResponse
+import spray.http.StatusCodes._
+import spray.routing.RequestContext
+import org.broadinstitute.dsde.firecloud.model.ModelJsonProtocol._
+import spray.httpx.SprayJsonSupport._
+
+import scala.collection.immutable.Iterable
+import scala.concurrent.Future
+import scala.util.{Failure, Success}
+
+object ProfileActor {
+  case class UpdateProfile(userInfo: UserInfo, profile: Profile)
+  def props(requestContext: RequestContext): Props = Props(new ProfileActor(requestContext))
+}
+
+class ProfileActor(requestContext: RequestContext) extends Actor with FireCloudRequestBuilding {
+
+  implicit val system = context.system
+  import system.dispatcher
+  val log = Logging(system, getClass)
+
+  override def receive: Receive = {
+
+    case UpdateProfile(userInfo: UserInfo, profile: Profile) =>
+      val pipeline = authHeaders(requestContext) ~> sendReceive
+      val profileProperties = ProfileValidator.propertyValueMap(profile)
+      val posts: Iterable[Future[HttpResponse]] = profileProperties map {
+        case (key, value) =>
+          val kv = FireCloudKeyValue(Some(key), Some(value))
+          pipeline {
+            Post(UserService.remoteSetKeyURL, ThurloeKeyValue(Some(userInfo.getUniqueId), Some(kv)))
+          }
+      }
+      val propertyUpdates: Future[List[HttpResponse]] = Future sequence posts.toList
+
+      val kv2 = FireCloudKeyValue(Some("isRegistrationComplete"), Some("true"))
+      val completionUpdate = pipeline {
+        Post(UserService.remoteSetKeyURL, ThurloeKeyValue(Some(userInfo.getUniqueId), Some(kv2)))
+      }
+
+      propertyUpdates onComplete {
+        case Success(responses) if responses.forall { _.status == OK } =>
+          log.debug("Profile is saved, update isRegistrationComplete status.")
+          completionUpdate onComplete {
+            case Success(response) if response.status == OK =>
+              context.parent ! RequestComplete(OK, response.toString)
+              context stop self
+            case Success(response) =>
+              context.parent ! RequestCompleteWithErrorReport(response.status, response.toString)
+              context stop self
+            case _ =>
+              context.parent ! RequestCompleteWithErrorReport(InternalServerError, "Failure posting profile: [%s]".format(profileProperties))
+              context stop self
+          }
+
+        case Success(responses) =>
```

Comment:
This block is also used in the entity-deletion code. It would be nice to extract this into a reusable utility or some Spray command (a route? a directive? I'm not even sure what the right object is). Don't hold up the PR for this; it's an optimization.